### PR TITLE
Fixed can't open epub in external app

### LINKED ISSUE #3453: Can't open EPUB from Project Gutenberg in external app
This reopens the issue from #2517. I see the same exact behavior. Either there was a regression, or this was never fixed. I have not tested PDFs because I don't see any PDFs in project gutenberg, only EPUBs, so this issue is only for EPUB.

#2517 was supposedly fixed by #2825. But that pr says it only fixes PDF, and says nothing about EPUB. So I'm guessing 2517 was never actually fixed, and was closed by mistake.

I don't see why the file type even matters. It should just write the file into the Download folder. Nonetheless, I have no problem opening EPUB files from other apps, like a file manager; they always open in my ebook reader (ReadEra) without issue.

**Environment**
- Version of Kiwix Android : 3.7.1
- Device : SM-N976U
- OS version : Android 11

### PR DESCRIPTION
**Pull request by MohitMaliFtechiz**
Fixes #3453 

* Added EPUB application viewing query to the manifest. This query provides access to check if an EPUB viewer application is installed on the user's device or not
* Added this query in `core/manifest` file so it will add this query to both `kiwix` and `custom` app.

https://github.com/kiwix/kiwix-android/assets/34593983/6867b2a3-5583-485d-841a-442034c87129

* To enhance the user experience, a condition has been implemented to check if a URI is a valid URL. If it is an external browser URL, it will be shown to the user; otherwise, the URI will not be displayed in the external link popup. This improvement ensured that only external browser URLs would be present to the user, providing a more user-friendly interface.

| With Epub Uri  | With Pdf Uri | With External browser URL |
| ------------- | ------------- | ------------- |
| epubUri |  pdfUri | validUrl |

* We are now showing proper error messages to the user when there is no app installed on their device to handle the pdf/epub/URL.

### CODE REVIEW

**Comment by dnut** [comment]
> What is a valid uri? There is no concept of valid or invalid link in Kiwix I'm aware of.

I think he's making a distinction between URL and URI. A URL is a type of URI. `content://...` is a URI because it identifies a resource, but you can argue that it is not a URL because it does not represent a network address for the resource.

This is only conjecture, but I think maybe the challenge is that the code that handles this...

> either an external link and then it should be open with external web browser or mail client (no other use case supported IMO).

... is generic. It is written to handle *any* URI that needs to be opened by android, regardless of whether it is a network address. It's android that figures out the URI needs to be opened by a web browser.

So this logic uses the same code:

> or internal link, and either ~the mime-type is supported by kiwix and then it is displayed~ or its not and then this is delegated to Android OS.

When it delegates to android, it uses the exact same logic as above that handles external links, which means you get an external link popup. This is how it was already implemented, and getting EPUBs working was just a matter of fixing an XML config file so that there was no error before it reaches that step.

In my mind, these are distinct bug fixes that require changes in completely different regions of code:
1. bugfix for EPUBs so they don't have an error.
2. Change the existing logic that asks android to open URIs so that it doesn't assume every URI is an external link. Instead it needs to parse the URI, and only display an "external links" popup if the URI represents a network address.

Maybe it's efficient to bundle both changes in a single PR, but it's worth considering the idea that either one could be deployed independently of the other, and each would only serve as an improvement.

>  If delegated to OS, a filepicker should open to be able to save it. If the OS can handle it (here via an epub reader), then it should be able to open it directly (so here the user should get a choice to download or open).

This feature would be nice to have. I think the existing logic is something like this:
1. automatically exports the file to the download folder without asking you to save it manually
2. constructs a URI representing the file and uses the generic logic to ask android to open the URI

But I agree that the ideal ux would allow:
- *any* mime type that kiwix cannot open may be stored locally in a file the user has access to, regardless of an intent being registered for that mime type, and regardless of any existing app on the device being able to open the file
- the user has the option between directly opening the file versus only saving the file into a folder
- if the file is only being saved, not opened, then the user can select which folder it is saved into

It's not clear to me how much code needs to change to make each of these possible.

**Comment by MohitMaliFtechiz** [comment]
> @MohitMaliFtechiz What is a valid uri? There is no concept of valid or invalid link in Kiwix I'm aware of.

@kelson42, I apologize for the delay in responding. I understand your concerns, and I want to provide clarity on the concept of valid or invalid URLs within Kiwix.

In our context, a valid URL refers to a link that is openable in an external browser. Conversely, an invalid URL represents a content URI that starts with the content:// prefix. These URIs are displayed when we attempt to open a pdf or epub file.

> Warnings are for external links, not for internal. Like written in the ticket I don't want to see them, they are inappropriate.

Yes, we are only showing warnings for external links, but here URIs are also external links(since they are not handled in the application) that's why they are showing in the external link dialog. I have only hidden these uri from the external link dialog because they look weird in the dialog as reported on the ticket(also  they are not openable in the external browser so if the user copies these URIs and pastes them in the browser they are not openable so it is unnecessary to show) 

> or internal link, and either the mime-type is supported by kiwix and then it is displayed or its not and then this is delegated to Android OS. If delegated to OS, a filepicker should open to be able to save it. If the OS can handle it (here via an epub reader), then it should be able to open it directly (so here the user should get a choice to download or open).

If it is an internal link then it is handled by Kiwix itself and if it is a mime-type that is not openable with OS e.g. image then we are showing a download media dialog that saves the image into storage.




And if it is handled by the OS, In the current scenario, we are downloading the file (epub, pdf) as soon as the user clicks on the pdf or epub file(download only once) and then we are showing an external link dialog if the user clicks on yes then it will be redirected to that app which handled that type of mime-type.
Here file picker in Android is for getting the file from the storage not for saving. However, if we only want to open the file in an external reader e.g. for epub so before opening the file it downloads that file and paas the downloaded file URI to the epub reader, so IMO, here giving a download option is not effective as it is already saving the file into the download folder.

**Comment by gouri-panda** [comment]
@MohitMaliDeveloper @kelson42 Is this PR ready?

**Comment by kelson42** [comment]
@gouri-panda please postpone review for the moment

**Comment by kelson42** [comment]
> In our context, a valid URL refers to a link that is openable in an external browser. Conversely, an invalid URL represents a content URI that starts with the content:// prefix. These URIs are displayed when we attempt to open a pdf or epub file.

Thank you for the clarification. I understand now, but the terminology is not the most appropriate. Here is how it should be:
* "invalid" URL means not respecting URL + ZIM URL standarts, this would be probabky a bug in the ZIM itself
* "dead" URL is what you mean: pointing to a resource which does not exist or is not found

Please correct the sourve code accordingly to this naming

**Comment by kelson42** [comment]
> Yes, we are only showing warnings for external links, but here URIs are also external links(since they are not handled in the application) that's why they are showing in the external link dialog.

Here we have an other vocabulary mix: "external" URL means that the URL points to a ressource which is external to the ZIM, typically somewhere on the internet. This has nothing to do with the mime-type of the ressource pointed and how it is handled within Kiwix. Please make code modification accordingly.

**Comment by kelson42** [comment]
@MohitMaliDeveloper I have left two comments 3 weeks ago, can you please handle them?

**Comment by MohitMaliFtechiz** [comment]
> @MohitMaliDeveloper I have left two comments 3 weeks ago, can you please handle them?

@kelson42 Sorry for being late on this, I will place a fix for it.

**Comment by kelson42** [comment]
@MohitMaliDeveloper Unfortunately this PR needs to be rebased (and fixing conflicts). Do you confirm as well this PR is not impacted somehow by #3501?

**Review by kelson42 — CHANGES_REQUESTED** [review]
@MohitMaliDeveloper My comments of https://github.com/kiwix/kiwix-android/pull/3455#issuecomment-1643567981 have not been taken in account.

File: core/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -22,6 +22,10 @@
     <intent>
       <action android:name="android.intent.action.TTS_SERVICE" />
     </intent>
+    <intent>
```
Comment (kelson42) [inline core/src/main/AndroidManifest.xml:25]: Please put a comment of the purpose of this? Why nothing about PDF?

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreWebViewClient.kt [hunk 2]
```diff
@@ -102,7 +103,11 @@ open class CoreWebViewClient(
             flags = Intent.FLAG_ACTIVITY_NO_HISTORY
             addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
           }
-          callback.openExternalUrl(intent)
+          val errorMessage = if (extension == "application/pdf")
+            R.string.no_pdf_application_installed
+          else
```
Comment (kelson42) [inline core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreWebViewClient.kt:108]: The whole function is wrong:
* Either this is an external link http/https and then you pass out to browser whatever
* Or this is an internal link and you pass to kiwix to open and then kiwix decides based on the mime-type stored in the ZIM to open itself or to pass to the OS to open via a third party app

File: core/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -63,6 +63,9 @@
   <string name="tts_lang_not_supported">The language of this page is not supported. The article may not be properly read.</string>
   <string name="no_reader_application_installed">Could not find an installed application for this type of file</string>
   <string name="no_email_application_installed">Please install an email service provider or email us at %1s</string>
+  <string name="no_browser_application_installed">Please install any web browser to open it e.g. chrome or firefox.</string>
+  <string name="no_pdf_application_installed">Please install any pdf reader application to open it.</string>
+  <string name="no_epub_application_installed">Please install any epub reader application to open it.</string>
```
Comment (kelson42) [inline core/src/main/res/values/strings.xml:68]: No mime-type specific error. Just something "Please install a reader application to be able to open this content."

**Review by kelson42 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**